### ISSUE TITLE
Display string instead of <span> element

### ISSUE THREAD

**Issue by guoyunhe** [formatjs/react-intl#987]
I need to translate something inside an `<option>` tag. If I put `<FormattedMessage />` component inside, I will got browser warnings. Because it output a `<span>` always.

Is there anyway to call some functions instead of using `<FormattedMessage />` component?

**Comment by ghostd** [comment]
Hi @guoyunhe 

You case uthe the 'tagName' property, see [the doc](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

**Comment by guoyunhe** [comment]
@ghostd If I use `tagName='option'` then I won't be able to add 'value' property to the '<option>' tag. What I want to get is:

```html
<option value='0'>Workday</option>
<option value='1'>Unpaid leave</option>
<option value='2'>Holiday</option>
```

**Comment by ghostd** [comment]
I've a waiting PR for your use case: #878 
Since it's not commented/merged, you could directly use the API with [formatMessage](https://github.com/yahoo/react-intl/wiki/API#formatmessage) and [injectIntl](https://github.com/yahoo/react-intl/wiki/API#injectintl)

**Comment by guoyunhe** [comment]
@ghostd thanks! The injectIntl works for me. Hope someday they have easier API functions to use.

**Comment by guangming21** [comment]
@guoyunhe

**Comment by mbrevda** [comment]
You can use a method prop to receive the formatted message and return any custom element:

(click image for live view)
[![image](https://user-images.githubusercontent.com/293004/31177763-abfb1b0e-a91f-11e7-90f2-1a8634f4299b.png)](https://www.webpackbin.com/bins/-KvbVdgqv_fJoOo6l06_)

**Comment by uberfabulous** [comment]
Update your App to use React 16.2.0+ and pass `textComponent={React.Fragment}` to wherever you call `<IntlProvider>`. E.g.: `<IntlProvider textComponent={React.Fragment}>...</IntlProvider>`.

**Comment by STRML** [comment]
We just updated our IntlProvider to use `textComponent={React.Fragment}` and it worked fantastically. One of those moments where you really appreciate the hard work so many teams work such that they interop flawlessly. Great work everyone.

**Comment by mbrevda** [comment]
I wonder if that shouldn't become the new default instead of span

**Comment by STRML** [comment]
It should if it exists (React >= 16.2.0).

**Comment by asual** [comment]
All works great with `textComponent={React.Fragment}` except the following warning which looks like an error in IE11 with `NODE_ENV=development`

```
Warning: React.createElement: type is invalid -- expected a string (for built-in components) 
or a class/function (for composite components) but got: object.

Check the render method of `FormattedMessage`.
    in FormattedMessage (created by Unknown)
    in h1 (created by Unknown)
    in div (created by Unknown)
    in Unknown (created by Unknown)
    in Unknown (created by Connect(Component))
```

**Comment by sergioviniciuss** [comment]
This won't work if you're using the FormattedHTMLMessage component though

**Comment by astorije** [comment]
Sorry if this is a stupid comment, I'm still _fairly_ new to the world of React. I just watched https://egghead.io/lessons/react-render-text-only-components-in-react-16 (1min), and it looks as though `Formatted*` are text-only, so they would fit the bill to return text directly in React 16.
Would that be an option?

**Comment by StorytellerCZ** [comment]
@astorije With React 16.2 it is possible to `React.Fragment` as a default instead of span, which in effect returns only text.

@mbrevda Given the backward compatibility to older versions of React, `react-intl` would have to make a major release to make this a new default which would also mean no longer supporting anything before React 16.2. I would be for that as it would allow to move on with other things as well.

**Comment by mbrevda** [comment]
Would you consider the absence of the expected `<span>` breaking?

**Comment by taion** [comment]
That would be a breaking change.

I think the next major bump should change the default tag, though.

**Comment by jquense** [comment]
post react 16 returning the plain string would also be fine, the `Rect.Fragment` is technically unnecessary if you want to return nothing but I see how this is more flexible 👍

**Comment by lauraseidler** [comment]
> This won't work if you're using the FormattedHTMLMessage component though

Since this is only due to `React.Fragment` not supporting `dangerouslySetInnerHTML` which only FormattedHTMLMessage needs (as far as I can see), this can be solved by using a custom component that only uses a span if FormattedHTMLMessage is used (if you're okay with the inconsistency):

```
const TranslationWrapper = ({ dangerouslySetInnerHTML, ...props }) =>
    dangerouslySetInnerHTML ? (
        <span dangerouslySetInnerHTML={dangerouslySetInnerHTML} {...props} />
    ) : (
        <React.Fragment {...props} />
    );
```

And then pass that into your Provider:
```
<IntlProvider textComponent={TranslationWrapper}>...</IntlProvider>
```

**Comment by agutoli** [comment]
<FormattedMessage id="title">
    {(txt) =>txt}
</FormattedMessage>

**Comment by agutoli** [comment]
`<FormattedMessage id="title">
    {(txt) => (
        <H1>
            {txt}
        </H1>
    )}
</FormattedMessage>`

**Comment by ZauberNerd** [comment]
> This won't work if you're using the FormattedHTMLMessage component though

You can use `<intlProvider textComponent={React.Fragment}>...</IntlProvider>` to globally switch to using Fragments and explicitly use a `span` for `FormattedHTMLMessage` like this:
```
<FormattedHTMLMessage tagName="span" />
```

**Comment by phifa** [comment]
Please change the default to "no-tag". Or at least make it possible with `<FormattedHTMLMessage tagName="" />`

**Comment by redonkulus** [comment]
I left a [comment in another issue](https://github.com/formatjs/react-intl/issues/999#issuecomment-493579362) about the possible costs to changing the default. I'm not against it and adding it to v3 would make sense, I just didn't want to cause unnecessary overhead for users when migrating. Of course, they could just set the textComponent field to `span` to avoid such issues. @DragonRaider5 also had a good option as well in that issue. 

I would prefer to continue the discussion in this issue as opposed to that one.

**Comment by longlho** [comment]
We cannot return `string` by default as that will break React Native. `React.Fragment` is the safer option w/o extra markup.

**Comment by STRML** [comment]
Good change in #1326. I still think we should support the empty string (`""`) which will do raw strings and avoid allocating the `React.Fragment`. for web.

**Comment by longlho** [comment]
@STRML 	mind making a new issue for it? Thanks! :)

**Comment by longlho** [comment]
closing in favor of #1330 and partially addressed by #1326
